src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -618,21 +618,45 @@ await foreach (ConfigurationSetting setting in client.GetConfigurationSettingsAs
                 }).ConfigureAwait(false);
             }
 
-                foreach (var loadOption in _options.KeyValueSelectors)
+            foreach (KeyValueSelector loadOption in _options.KeyValueSelectors)
+            {
+                IAsyncEnumerable<ConfigurationSetting> settingsEnumerable;
+
+                if (string.IsNullOrEmpty(loadOption.SnapshotName))
+                {
+                    settingsEnumerable = client.GetConfigurationSettingsAsync(
+                        new SettingSelector
+                        {
+                            KeyFilter = loadOption.KeyFilter,
+                            LabelFilter = loadOption.LabelFilter
+                        },
+                        cancellationToken);
+                }
+                else
                 {
-                    var selector = new SettingSelector
+                    ConfigurationSettingsSnapshot snapshot = await client.GetSnapshotAsync(loadOption.SnapshotName);
+
+                    if (snapshot != null)
                     {
-                        KeyFilter = loadOption.KeyFilter,
-                        LabelFilter = loadOption.LabelFilter
-                    };
+                        if (snapshot.CompositionType == null || !snapshot.CompositionType.ToString().Equals("key"))
+                        {
+                            throw new ArgumentException($"{nameof(snapshot.CompositionType)} for the selected snapshot must be 'key', found '{snapshot.CompositionType}'.");
+                        }
+                    }
+
+                    settingsEnumerable = client.GetConfigurationSettingsForSnapshotAsync(

```

Comment:
What are our thoughts about caching the snapshot results so we don't reload it during refresh?

Reply:
The only thing that comes to mind is we'd need to be able to tell if a snapshot was deleted and one with the same name was created with different key-values. Maybe we could see that it's a different snapshot based on the time it was created?

Reply:
I'm a little torn on the idea. 

On one hand snapshots are a perfect candidate for caching because they are mostly static.

However, an app setup with refresh is already establishing a dependency on querying key-values during runtime, so it waters down the value of caching snapshots for me a little bit. 

The downside of the caching would be added memory pressure of caching the snapshot response kvs. The previous comment about refresh deters me from suggesting adding more things in memory.

Reply:
When refresh is enabled, we will need the data of snapshots anyway, so I won't worry about the memory if we cache them. It doesn't make sense to reload the snapshots every few seconds. We can skip caching if the refresh is not enabled.

It's an interesting scenario Amer brought up. The snapshot is not expected to change, but it could change in the scenario you mentioned. I would rather cache it, so the configuration won't change due to a refresh.

Reply:
> It doesn't make sense to reload the snapshots every few seconds.

Is there a scenario that would cause snapshots to be reloaded every few seconds? The only way I would imagine them being reloaded is with sentinel key refresh.

Reply:
Sure, you are right. Reloading the snapshot could be less frequent. Everything else stays true.

Reply:
> I'm not sure why would refresh change what we are storing for snapshots?

Not sure I follow the question. When a sentinel key is changed and we need to reload all, don't we need to reload the snapshot if it's not cached?

Reply:
Discussed offline with Zhenlan. 

The main scenario is sentinel key refresh
* Snapshots don't play a role in detection of sentinel key change
* A reload all operation would take a service dependency on more than snapshots, otherwise refresh wouldn't be configured. This waters down the value of snapshot caching.

We agreed that while there could be a perceived benefit of snapshot caching by reducing requests to the service during a refresh, it doesn't seem like the benefit is enough to justify the memory utilization for caching.